util/patch_solr78_bytecode.py

```diff
@@ -0,0 +1,153 @@
+    this function will patch the resulting byte code to use it as
+    an abstract base class instead to be compatible with Solr<9.
+    """
+    buf = bytearray(bytecode)
+    if buf[:4] != b"\xca\xfe\xba\xbe":
+        raise ValueError("Not a classfile, corrupt JAR?")
+    constant_pool_count = struct.unpack_from(">H", buf, 8)[0]
+
+    # Traverse constant pool
+    constant_pool_size = 0
+    constant_pool_start = 10
+    for _ in range(constant_pool_count - 1):
+        tag = buf[constant_pool_start + constant_pool_size]
+        # Modify reference to super class constructor to point to our
+        # new base class instead of java.lang.Object<init>
+        if tag == 10:  # method reference
+            target_idx = constant_pool_start + constant_pool_size + 2
+            if buf[target_idx] == 7:  # reference to java.lang.Object
+                buf[target_idx] = 8  # change to CloseHook
+        # Update the constant pool size
+        if isinstance(CONSTANT_POOL_SIZES[tag], int):
+            constant_pool_size += CONSTANT_POOL_SIZES[tag]
+        else:
+            constant_pool_size += CONSTANT_POOL_SIZES[tag](
+                buf[constant_pool_start + constant_pool_size :]
+            )
+    after_pool_idx = constant_pool_start + constant_pool_size
+    iface_count = struct.unpack_from(">H", buf, after_pool_idx + 6)[0]
+    if iface_count != 1:
+        return bytes
+
+    # Start of interface pointers
+    ifaces_idx = after_pool_idx + 8
+    # Object pool index of the `CloseHook` type
+    closehook_idx = struct.unpack_from(">H", buf, ifaces_idx)[0]
+    # Strip interfaces list from class file
+    del buf[ifaces_idx : ifaces_idx + 2]
+    # Set `CloseHook` as the super class
+    buf[after_pool_idx + 4 : after_pool_idx + 6] = struct.pack(">H", closehook_idx)
+    # Clear all interfaces for the class
+    buf[after_pool_idx + 6 : after_pool_idx + 8] = b"\x00\x00"
+    return bytes(buf)
+
+
+def patch_jar(source_path: Path, target: BinaryIO) -> None:
+    with zipfile.ZipFile(source_path) as source_jar, zipfile.ZipFile(
+        target, "w", compression=zipfile.ZIP_DEFLATED
+    ) as target_jar:
+        for fname in source_jar.namelist():
+            data = source_jar.read(fname)
+            if fname.endswith(".class"):
+                data = patch_package_paths(data)
+            if fname.endswith("OcrHighlightComponent$1.class"):
+                data = patch_close_hook(data)
+            target_jar.writestr(fname, data)
+
+
+if __name__ == "__main__":
+    if len(sys.argv) < 2:
+        print(__doc__)
+        print("Usage:")
+        print("$ patch_solr78_bytecode.py <source_jar> [<target_jar>]", file=sys.stderr)
+        sys.exit(1)
+    source_path = Path(sys.argv[1])
+    if not source_path.exists():
+        print(f"File at {source_path} does not exist!")
+        sys.exit(1)
+    if len(sys.argv) == 3:
+        target_path = Path(sys.argv[2])
+        if not source_path.exists():
```

Comment:
Check sollte in Zeile 142 schon gemacht sein :thinking:

Reply:
Nein, das ist ein anderer Check.

- `len(sys.argv) == 3` ->  User hat Ziel-Pfad direkt angegeben, dorthin schreiben
- `len(sys.argv) == 2` -> User hat nur Quell-Pfad angegeben, nach `stdout` schreiben